src/main/java/io/smartcat/ranger/core/parser/DataGenerator.java

```diff
@@ -0,0 +1,117 @@
+package io.smartcat.ranger.core.parser;
+
+import io.smartcat.ranger.core.CompositeValue;
+import io.smartcat.ranger.core.PrimitiveValue;
+import io.smartcat.ranger.core.Value;
+import io.smartcat.ranger.core.ValueProxy;
+import org.parboiled.Parboiled;
+import org.parboiled.parserunners.ReportingParseRunner;
+import org.parboiled.support.ParsingResult;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Generates data specified by configuration.
+ */
+public class DataGenerator {
+
+    private CompositeValue outputValue;
+    private Map<String, ValueProxy<?>> proxyValues;
+
+    private DataGenerator(Map<String, ValueProxy<?>> proxyValues, String outputValueName) {
+        Value<?> val = proxyValues.get(outputValueName);
+        if (val instanceof CompositeValue) {
+            this.outputValue = (CompositeValue) val;
+        } else {
+            this.outputValue = new CompositeValue();
+            this.outputValue.setValue("output", val);
+        }
+        this.proxyValues = proxyValues;
+    }
+
+    /**
+     * Returns next value.
+     *
+     * @return Next value.
+     */
+    public Map<String, Object> next() {
+        Map<String, Object> output = outputValue.eval();
+        reset();
+        return output;
+    }
+
+    private void reset() {
+        for (ValueProxy<?> proxy : proxyValues.values()) {
+            proxy.reset();
+        }
+    }
+
+    /**
+     * Builds {@link DataGenerator} based on specified <code>config</code> and <code>outputValueName</code>.
+     *
+     * @param config Data generation configuration.
+     * @param outputValueName Name of the output value.
+     * @return An instance of {@link DataGenerator}.
+     */
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    public static DataGenerator build(Map<String, Object> config, String outputValueName) {
+        Map<String, ValueProxy<?>> proxyValues = new HashMap<>();
+        ValueExpressionParser parser = Parboiled.createParser(ValueExpressionParser.class);
+        parser.setProxyValues(proxyValues);
+        ReportingParseRunner<Value<?>> parseRunner = new ReportingParseRunner<>(parser.value());
+
+        for (Map.Entry<String, Object> entry : config.entrySet()) {
+            String valueName = entry.getKey();
+            Object valueDef = entry.getValue();
+            Value<?> value = parse(valueDef, parseRunner, proxyValues);
+            ValueProxy proxy = obtainProxy(valueName, proxyValues);
+            proxy.setDelegate(value);
```

Comment:
Why is it important to obtain first proxy and then set delegate. If we refactor Proxy to force constructor with delegate we can always construct new proxy and use put with same key, this way we will either add new to map or replace existing.